### PR TITLE
Fixing error in PartialDerivative operator for tensor valence

### PR DESCRIPTION
**Pull request by Upabjojr**
#### Release Notes

* tensor
  * Fix PartialDerivative operator to correctly handle the valence (covariant/contravariant) of the deriving variable.

### CODE REVIEW

File: sympy/tensor/toperators.py
Comment (Upabjojr) [inline sympy/tensor/toperators.py]: @joha2 I have fixed it like this. `PartialDerivative(A(i), A(i))` gets printed fully-contravariant. When you access the indices, one is contravariant, while the deriving index gets flipped into covariant.

Indeed, deriving by a contravariant symbol adds a covariant slot to the resulting tensor.

File: sympy/tensor/tests/test_tensor_operators.py
Comment (Upabjojr) [inline sympy/tensor/tests/test_tensor_operators.py]: @joha2 these are examples of replacing the tensorial expression with arrays (sometimes you may wish to switch to the array expression). Do you think they are correct?

Comment (joha2) [reply sympy/tensor/tests/test_tensor_operators.py]: @Upabjojr for the example in line 57 I am not sure, since you derive a contravariant tensor by its covariant components, at least there should arise some metric components. In line 60 on the other hand there should arise strictly the Kronecker delta, independently of the metric. So maybe for a thorough test, one should use a non-trivial metric to distuingish these two cases.

Comment (joha2) [reply sympy/tensor/tests/test_tensor_operators.py]: All other (abstract) tests are okay.

File: sympy/tensor/toperators.py
Comment (joha2) [reply sympy/tensor/toperators.py]: I think as long as the pretty print is okay, people can live with the non intuitive output of a non-performed `PartialDerivative` :-)

**Comment by joha2** [comment]
@Upabjojr is there anything left to do for me, like performing the actual review? I never did this before. Or could this PR be merged without my interaction?

**Comment by Upabjojr** [comment]
> @Upabjojr is there anything left to do for me, like performing the actual review? I never did this before. Or could this PR be merged without my interaction?

If you want to review, you're welcome to do it. I would like if someone can check my code.

I think we need to integrate your code into this one. But maybe better in a separate PR.

File: sympy/tensor/toperators.py
Comment (Upabjojr) [reply sympy/tensor/toperators.py]: I have also pushed the LaTeX printers, so people should be able to see a nice output in Jupyter notebooks.

File: sympy/tensor/tests/test_tensor_operators.py
Comment (Upabjojr) [reply sympy/tensor/tests/test_tensor_operators.py]: You're right. This is probably wrong.

Comment (Upabjojr) [reply sympy/tensor/tests/test_tensor_operators.py]: > for the example in line 57 I am not sure, since you derive a contravariant tensor by its covariant components, at least there should arise some metric components.

The metric is passed with `L: diag( ... )`. The operator `diag( ... )` creates a diagonal matrix.

**Comment by joha2** [comment]
Okay, how should it be done?

Is this the right way to go?
* I close my PR #18093 
* We wait for your `PartialDerivative` changes to be merged
* Then I will merge `TensPartialDerivative` into `PartialDerivative` and submit a new PR

Or this there another possibility, where my PR does not have to be closed?

**Comment by Upabjojr** [comment]
> Or this there another possibility, where my PR does not have to be closed?

Well... there is no big difference. You can open as many PRs as you like, but in most cases it makes no sense. You can also merge this branch into yours.

**Comment by Upabjojr** [comment]
@joha2 I made an update to the tests. The results were correct earlier (the code was performing raising/lowering operations after the derivative was computed, I've removed that part).

If you don't see any mistake, we can merge this PR.

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

File: sympy/tensor/tests/test_tensor_operators.py
Comment (joha2) [inline sympy/tensor/tests/test_tensor_operators.py]: OK

**Review by joha2 — APPROVED** [review]
The tests are all fine.

**Comment by Upabjojr** [comment]
> The tests are all fine.

Need to understand why they are not fine in Python 2.7 and 3.5

**Comment by joha2** [comment]
Strange that the free indices need to be sorted before. Why is this the case?

**Comment by Upabjojr** [comment]
> Strange that the free indices need to be sorted before. Why is this the case?

They are probably read from a `dict` dictionary or a `set`. The way elements are sorted in a dictionary is quite random and may vary between Python versions.

**Comment by joha2** [comment]
@Upabjojr once your branch is merged into master, I will do `git pull upstream master` for my branch. Is this the correct way to go? Do I have to rebase afterwards to not clutter the commit messages by merges?

**Comment by joha2** [comment]
Another question comes to my mind:

In the line:
https://github.com/sympy/sympy/blob/1797bd5d185fe59cf5783eb3a87f17eafff198f5/sympy/tensor/toperators.py#L44

Why would one verify that all variables have rank one? One can also perfectly derive by a multi index tensor:
g3341

Or did I get this TODO wrong?

**Comment by Upabjojr** [comment]
I usually check out into master, pull, check out the branch again and then merge master. I don't know if it's the best way.

Concerning the merging, you can choose how to do it.

**Comment by Upabjojr** [comment]
Indeed there are cases in which you'd derive by a matrix.

**Comment by Upabjojr** [comment]
I would also prefer to implement your code in a different way: maybe only Tensor classes should be derived.

The remainder could be handled as expansion rules in doit()

**Comment by joha2** [comment]
> I usually check out into master, pull, check out the branch again and then merge master. I don't know if it's the best way.

I think I will also do it that way.

> I would also prefer to implement your code in a different way: maybe only Tensor classes should be derived.
> The remainder could be handled as expansion rules in doit()

I am not sure whether I understand correctly: Is the problem the mixup of `diff` and `_eval_partial_derivatives` as well as the `isinstance` checks? Could you more elaborate on that? I also wanted the possibility to derive by scalars in scalar expressions.

**Comment by Upabjojr** [comment]
What I mean is to separate the derivative into two different steps:

1. Expand the expression to be derived. For example `PartialDerivative(a*b, c) ==> PartialDerivative(a, c)*b + a*PartialDerivative(b, c)`.
2. Compute the derivative. At this point you only have partial derivatives of tensors (all TensMul and TensAdd should have been expanded).

Steps 1 and 2 may be called by two different methods.

**Comment by joha2** [comment]
Yes, that is a good idea. I also had something like this in mind, when I wrote about the error appearing in the tests in my PR #18093. So this means, `doit()` should just expand to maintain the tensor structure (i.e. to be compatible with usage of `*` which means `TensMul.doit()` in contrast to just `TensMul`) and the second step could be and `evaluate()` method. For sake of fast evaluation, it would be nice to have a `dependency` list in the tensors. I.e. if `H(i, j)` has dependency `A`, this means `PartialDerivative(H(i, j), A(k))` stays as it is, but `PartialDerivative(H(i, j), B(k))` vanishes. Is this useful? Btw. at the end of the day not only Tensors are in `PartialDerivative`, there is also the possibility that there are symbols. This edge case also has to be considered:
* `PartialDerivative(tau*A(i), B(j)) = PartialDerivative(tau, B(j))*A(i) + PartialDerivative(A(i), B(j))*tau`
* `PartialDerivative(tau*A(i), tau) = PartialDerivative(tau, tau)*A(i) + tau*PartialDerivative(A(i), tau)`

**Comment by Upabjojr** [comment]
Yes, though I would recommend two different private methods,  so we can call then separately.

**Pull request merged**
